modules/lsp.nix

```
@@ -0,0 +1,96 @@
+{
+  lib,
+  config,
+  ...
+}:
+let
+  inherit (lib) mkOption types;
+
+  cfg = config.lsp;
+
+  serverType = types.submodule {
+    options = {
+      enable = lib.mkEnableOption "the language server";
+
+      config = mkOption {
+        type = with types; attrsOf anything;
+        description = ''
+          Configurations for each language server.
+        '';
+        default = { };
+        example = {
+          cmd = [
+            "clangd"
+            "--background-index"
+          ];
+          root_markers = [
+            "compile_commands.json"
+            "compile_flags.txt"
+          ];
+          filetypes = [
+            "c"
+            "cpp"
+          ];
+        };
+      };
+    };
+  };
+in
+{
+  options.lsp = {
+    servers = mkOption {
+      type = types.attrsOf serverType;
+      description = ''
+        Wrapper of the `vim.lsp` API.
+      '';
+      default = { };
+      example = {
+        "*".config = {
+          root_markers = [ ".git" ];
+          capabilities.textDocument.semanticTokens = {
+            multilineTokenSupport = true;
+          };
+        };
+        luals.enable = true;
+        clangd = {
+          enable = true;
+          config = {
+            cmd = [
+              "clangd"
+              "--background-index"
+            ];
+            root_markers = [
+              "compile_commands.json"
+              "compile_flags.txt"
+            ];
+            filetypes = [
+              "c"
+              "cpp"
+            ];
+          };
+        };
+      };
+    };
+  };
+
+  config = {
+    extraConfigLua =
+      let
+        serversToEnable = lib.pipe cfg.servers [
+          (lib.filterAttrs (_: props: props.enable))
+          (lib.mapAttrsToList (serverName: _: serverName))
+        ];
+      in
+      ''
+        do
+          local __nixvimLspServersConfigs = ${lib.nixvim.toLuaObject cfg.servers}
+
+          for server_name, server_props in ipairs(__nixvimLspServersConfigs) do
+            vim.lsp.config[server_name] = server_props.config
+          end
```

```suggestion
                    for server_name, server_props in pairs(__nixvimLspServersConfigs) do
            if server_props.config then
              vim.lsp.config[server_name] = server_props.config
            end
          end
```

---

Alternatively, we could filter out empty configs when writing `__nixvimLspServersConfigs`?

Might be slightly more efficient by moving the logic from runtime to buildtime.

---

```suggestion
          for server_name, server_props in ipairs(__nixvimLspServersConfigs) do
            vim.lsp.config(server_name, server_props.config)
          end
```

IMO `__nixvimLspServersConfigs` should also be `mapAttrs (_: props: props.config)` so that we don't end up needlessly serialising the other props (like `enable`).

I also wonder if it makes sense that we are setting the config for _all_ defined servers, even ones with `enable = false`?

I know we're using `enable` later when deciding which servers to, well, _enable_; I just wonder if it's confusing that disabled servers still get configured?

---

> > IMO `__nixvimLspServersConfigs` should also be `mapAttrs (_: props: props.config)`, so that we don't end up needlessly serialising the unrelated non-config props (like `enable`).
> 
> But, how do you get `server_name` then?

The name would be unchanged (that's the `_` in the `mapAttrs`). However I personally think https://github.com/nix-community/nixvim/pull/3203#discussion_r2055109527 is a better idea than this one.

> The user could want to enable the server somewhere else (in an autocmd for instance) but still have it configured through `lsp.servers.foo.config`.
> As the API itself differentiates configuring and enabling servers, I think that our wrapping should do the same.

Yeah, I was on the fence regarding the same argument. IMO this could be a justification for a global `lsp.enable` option, even if it is default-true, just to allow users to disable (e.g. conditionally) us configuring stuff for them